tdigest.go

```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
 		other.Add(c.mean, c.count)
 	}
 }
+
+func (cs *centroidSet) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	w := &stickyErrWriter{w: buf, err: nil}
+	binary.Write(w, binary.LittleEndian, int32(1))
+
+	var min, max float64
+	if len(cs.centroids) > 0 {
+		min = cs.centroids[0].mean
+		max = cs.centroids[len(cs.centroids)-1].mean
+	}
+
+	binary.Write(w, binary.LittleEndian, min)
+	binary.Write(w, binary.LittleEndian, max)
+	binary.Write(w, binary.LittleEndian, float64(cs.compression))
+	binary.Write(w, binary.LittleEndian, int32(len(cs.centroids)))
+	if w.err != nil {
+		return nil, w.err
+	}
+	for _, c := range cs.centroids {
+		binary.Write(w, binary.LittleEndian, int64(c.count))
+		binary.Write(w, binary.LittleEndian, float64(c.mean))
+	}
+	if w.err != nil {
+		return nil, w.err
+	}
+	return buf.Bytes(), nil
+}
+
+func (cs *centroidSet) UnmarshalBinary(p []byte) error {
+	var (
+		encodingVersion int32
+		min, max        float64
+		n               int32
+	)
+	r := &stickyErrReader{r: bytes.NewReader(p), err: nil}
+
+	binary.Read(r, binary.LittleEndian, &encodingVersion)
+	if encodingVersion != 1 {
+		return fmt.Errorf("invalid encoding version: %d", encodingVersion)
+	}
+
+	binary.Read(r, binary.LittleEndian, &min)
+	binary.Read(r, binary.LittleEndian, &max)
+	binary.Read(r, binary.LittleEndian, &cs.compression)
+	binary.Read(r, binary.LittleEndian, &n)
+	if r.err != nil {
+		return r.err
+	}
+
+	cs.centroids = make([]*centroid, int(n))
+	for i := 0; i < int(n); i++ {
+		var (
+			count int64
+			mean  float64
+		)
+		binary.Read(r, binary.LittleEndian, &count)
+		binary.Read(r, binary.LittleEndian, &mean)
+		if r.err != nil {
+			return r.err
+		}
+		cs.centroids[i] = &centroid{count: int(count), mean: mean}
```

Comment:
Does this package have invariants that can be broken by unmarshaling bad/crafted input? If those can lead to a panic later (I haven't examined that), we should check the invariants here and return an error.

Reply:
Maybe, but I don't think so. I'll try fuzzing.